sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -156,29 +157,36 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         case lit @ Literal(null, _) => nullList += lit
         case lit @ NonNullLiteral(_, _) =>
           unwrapCast(EqualTo(in.value, lit)) match {
-            case EqualTo(_, unwrapLit: Literal) => canCastList += unwrapLit
-            case e @ And(IsNull(_), Literal(null, BooleanType)) => cannotCastList += e
+            // the function `unwrapCast` returns None means the literal can not cast to fromType,
+            // for instance: (the boundreference is of type DECIMAL(5,2))
+            //     CAST(boundreference() AS DECIMAL(10,4)) = 123456.1234BD
+            // Due to `cast(lit, fromExp.dataType) == null` we simply return
+            // `falseIfNotNull(fromExp)`.
+            case None | Some(And(IsNull(_), Literal(null, BooleanType))) =>
```

Comment:
what's wrong with `And(IsNull(_), Literal(null, BooleanType))`?

Reply:
As the comment shown in L149 ~ L153

> There are 3 kinds of literals in the in.list:
> 1. null literals
> 2. The literals that can cast to fromExp.dataType
> 3. The literals that cannot cast to fromExp.dataType

for 3, we have tow cases.

- A. the literal cannot cast to fromExp.dataType, and there is no min/max for the `fromType`, then `unwrapCast` returns None

for instance:

```
cast(input[2, decimal(5,2), true] as decimal(10,4)) = 123456.1234
```


- B. the literal `value` is out of `fromType` range `(min, max)`, then `unwrapCast` returns `falseIfNotNull(fromExp)`

for instance:

```
cast(input[0, smallint, true] as bigint) = 2147483647
```

Reply:
Actually, I missed another case before.
- C. the literal value is rounded up/down after casting to `fromType`, for instance:
```
cast(input[1, float, true] as double) = 3.14
```
 note that 3.14 will be rounded to 3.14000010... after casting to float